File: src/22/day22.py
from collections import namedtuple

Brick = namedtuple('Brick', ['x1', 'x2', 'y1', 'y2', 'z1', 'z2'])

# `bricks` and `lowest_heights` will be mutated
def pushed_height(
        index: int,
        dependencies: list[list[int]],
        bricks: list[Brick],
        lowest_heights: dict[int, int]) -> int:
    if (result := lowest_heights.get(index, None)) is not None:
        return result
    if dependencies[index]:
        lowest_z = max(
            pushed_height(i, dependencies, bricks, lowest_heights)
            for i in dependencies[index]
        ) + 1
        height = bricks[index].z2 - bricks[index].z1
        bricks[index] = Brick(
            bricks[index].x1, bricks[index].x2,
            bricks[index].y1, bricks[index].y2,
            lowest_z, lowest_z + height
        )
        lowest_heights[index] = lowest_z + height
        return lowest_z + height
    else:
        height = bricks[index].z2 - bricks[index].z1
        bricks[index] = Brick(
            bricks[index].x1, bricks[index].x2,
            bricks[index].y1, bricks[index].y2,
            0, height
        )
        lowest_heights[index] = height
        return height

def part1(filename: str) -> int:
    bricks = []
    with open(filename, 'rt') as f:
        for line in f:
            e1, e2 = line.strip().split('~')
            x1, y1, z1 = [int(x) for x in e1.split(',')]
            x2, y2, z2 = [int(x) for x in e2.split(',')]
            bricks.append(
                Brick(
                    min(x1, x2), max(x1, x2),
                    min(y1, y2), max(y1, y2),
                    min(z1, z2), max(z1, z2)
                )
            )
    dependencies = [[] for _ in range(len(bricks))]
    for i in range(len(bricks)):
        this = bricks[i]
        for j in range(len(bricks)):
            if i == j:
                continue
            that = bricks[j]
            if (that.z2 < this.z1 and
                that.x2 >= this.x1 and that.x1 <= this.x2 and
                that.y2 >= this.y1 and that.y1 <= this.y2):
                dependencies[i].append(j)
    lowest_heights = dict()
    for i in range(len(bricks)):
        pushed_height(i, dependencies, bricks, lowest_heights)
    dependencies = [
        [dep for dep in dependencies[i] if bricks[dep].z2 + 1 == bricks[i].z1]
        for i in range(len(bricks))
    ]
    return len(bricks) - len({s[0] for s in dependencies if len(s) == 1})

def part2(filename: str) -> int:
    bricks = []
    with open(filename, 'rt') as f:
        for line in f:
            e1, e2 = line.strip().split('~')
            x1, y1, z1 = [int(x) for x in e1.split(',')]
            x2, y2, z2 = [int(x) for x in e2.split(',')]
            bricks.append(
                Brick(
                    min(x1, x2), max(x1, x2),
                    min(y1, y2), max(y1, y2),
                    min(z1, z2), max(z1, z2)
                )
            )
    dependencies = [[] for _ in range(len(bricks))]
    for i in range(len(bricks)):
        this = bricks[i]
        for j in range(len(bricks)):
            if i == j:
                continue
            that = bricks[j]
            if (that.z2 < this.z1 and
                that.x2 >= this.x1 and that.x1 <= this.x2 and
                that.y2 >= this.y1 and that.y1 <= this.y2):
                dependencies[i].append(j)
    lowest_heights = dict()
    for i in range(len(bricks)):
        pushed_height(i, dependencies, bricks, lowest_heights)
    dependencies = [
        [dep for dep in dependencies[i] if bricks[dep].z2 + 1 == bricks[i].z1]
        for i in range(len(bricks))
    ]
    s = 0
    for i in range(len(bricks)):
        falling = {i}
        k = 0
        while True:
            l = len(falling)
            new_entries = [
                j for j in range(len(bricks))
                if all(x in falling for x in dependencies[j])
                and len(dependencies[j]) > 0
            ]
            for e in new_entries: falling.add(e)
            if len(falling) == l:
                break
        s += len(falling) - 1
    return s

if __name__ == '__main__':
    print(part1('day22_input.txt'))
    print(part2('day22_input.txt'))

File: src/22/day22.rs
use std::collections::{BTreeSet, HashMap};

pub struct Brick {
    pub x1: u32,
    pub x2: u32,
    pub y1: u32,
    pub y2: u32,
    pub z1: u32,
    pub z2: u32
}

fn pushed_height(index: usize,
                 dependencies: &Vec<Vec<usize>>,
                 bricks: &mut Vec<Brick>,
                 lowest_heights: &mut HashMap<usize, u32>) -> u32 {
    if let Some(&result) = lowest_heights.get(&index) {
        return result;
    }

    let lowest_z = if !dependencies[index].is_empty() {
        dependencies[index]
        .iter()
        .map(|&i| pushed_height(i, dependencies, bricks, lowest_heights))
        .reduce(|acc, x| if acc > x { acc } else { x })
        .unwrap()
        + 1
    } else {
        0
    };

    let height = bricks[index].z2 - bricks[index].z1;
    bricks[index].z1 = lowest_z;
    bricks[index].z2 = lowest_z + height;
    lowest_heights.insert(index, lowest_z + height);

    return lowest_z + height;
}

fn part1(filename: &str) -> usize {
    let lines = std::fs::read_to_string(filename).unwrap();
    let mut bricks = lines
        .trim()
        .split('\n')
        .map(|x| x.trim().split('~'))
        .map(
            |mut x| (
            x.next().unwrap().split(',').map(|x| x.parse::<u32>().unwrap()).collect::<Vec<_>>(),
            x.next().unwrap().split(',').map(|x| x.parse::<u32>().unwrap()).collect::<Vec<_>>()
            )
        )
        .map(
            |(p1, p2)|
            Brick {
                x1: std::cmp::min(p1[0], p2[0]),
                x2: std::cmp::max(p1[0], p2[0]),
                y1: std::cmp::min(p1[1], p2[1]),
                y2: std::cmp::max(p1[1], p2[1]),
                z1: std::cmp::min(p1[2], p2[2]),
                z2: std::cmp::max(p1[2], p2[2])
            }
        )
        .collect::<Vec<_>>();
    let mut dependencies = (0..bricks.len())
        .map(|_| Vec::<usize>::new())
        .collect::<Vec<_>>();
    for i in 0..bricks.len() {
        for j in 0..bricks.len() {
            if i == j { continue; }
            if bricks[j].z2 < bricks[i].z1
            && bricks[j].x2 >= bricks[i].x1 && bricks[j].x1 <= bricks[i].x2
            && bricks[j].y2 >= bricks[i].y1 && bricks[j].y1 <= bricks[i].y2 {
                dependencies[i].push(j);
            }
        }
    }
    let mut lowest_heights = HashMap::<usize, u32>::new();
    for i in 0..bricks.len() {
        pushed_height(i, &dependencies, &mut bricks, &mut lowest_heights);
    }
    dependencies = (0..bricks.len())
        .map(
            |i|
            dependencies[i]
            .iter()
            .filter(|&&dep| bricks[dep].z2 + 1 == bricks[i].z1)
            .cloned()
            .collect()
        )
        .collect();
    let mut single_deps = Vec::<usize>::new();
    for s in dependencies {
        if s.len() == 1 && !single_deps.contains(&s[0]) {
            single_deps.push(s[0]);
        }
    }
    return bricks.len() - single_deps.len();
}

fn part2(filename: &str) -> usize {
    let lines = std::fs::read_to_string(filename).unwrap();
    let mut bricks = lines
        .trim()
        .split('\n')
        .map(|x| x.trim().split('~'))
        .map(
            |mut x| (
            x.next().unwrap().split(',').map(|x| x.parse::<u32>().unwrap()).collect::<Vec<_>>(),
            x.next().unwrap().split(',').map(|x| x.parse::<u32>().unwrap()).collect::<Vec<_>>()
            )
        )
        .map(
            |(p1, p2)|
            Brick {
                x1: std::cmp::min(p1[0], p2[0]),
                x2: std::cmp::max(p1[0], p2[0]),
                y1: std::cmp::min(p1[1], p2[1]),
                y2: std::cmp::max(p1[1], p2[1]),
                z1: std::cmp::min(p1[2], p2[2]),
                z2: std::cmp::max(p1[2], p2[2])
            }
        )
        .collect::<Vec<_>>();
    let mut dependencies = (0..bricks.len())
        .map(|_| Vec::<usize>::new())
        .collect::<Vec<_>>();
    for i in 0..bricks.len() {
        for j in 0..bricks.len() {
            if i == j { continue; }
            if bricks[j].z2 < bricks[i].z1
            && bricks[j].x2 >= bricks[i].x1 && bricks[j].x1 <= bricks[i].x2
            && bricks[j].y2 >= bricks[i].y1 && bricks[j].y1 <= bricks[i].y2 {
                dependencies[i].push(j);
            }
        }
    }
    let mut lowest_heights = HashMap::<usize, u32>::new();
    for i in 0..bricks.len() {
        pushed_height(i, &dependencies, &mut bricks, &mut lowest_heights);
    }
    dependencies = (0..bricks.len())
        .map(
            |i|
            dependencies[i]
            .iter()
            .filter(|&&dep| bricks[dep].z2 + 1 == bricks[i].z1)
            .cloned()
            .collect()
        )
        .collect();
    let mut s = 0;
    for i in 0..bricks.len() {
        let mut falling = BTreeSet::<usize>::new();
        falling.insert(i);
        loop {
            let l = falling.len();
            for j in 0..bricks.len() {
                if !dependencies[j].is_empty() && dependencies[j].iter().all(|x| falling.contains(&x)) {
                    falling.insert(j);
                }
            }
            if falling.len() == l { break; }
        }
        s += falling.len() - 1;
    }
    return s;
}

fn main() {
    println!("{}", part1("day22_input.txt"));
    println!("{}", part2("day22_input.txt"));
}
